### PR TITLE
[Relay] WithFields method for Call, Function, Var, TupleGetItem, If, Let, RefCreate, RefRead, RefWrite, Match, and Clause

### PR DESCRIPTION
**Pull request by electriclilies**
In this PR, I implement the WithFields method for Call, Function, Var, TupleGetItem, If, Let, RefCreate, RefRead, RefWrite, Match and Clause. It is a continuation of work in #9533, where I implement WithFields for Tuples.  
Additionally, I remove the explicit COW logic from the visitors in expr_functor.cc for these classes. In future work, I'll use these methods to replace most explicit reconstructions of these classes in visitors.
@mbs-octoml @mikepapadim PTAL! I would especially appreciate typo checks in the documentation, I did a lot of copy-pasting and probably missed some things there.
cc @tqchen in case you are interested -- I've kept the API consistent with what we discussed for WithFields for tuples

### CODE REVIEW

**Review by mikepapadim — COMMENTED** [review]
LGTM, maybe add a `cpp` tests for new Clause.

File: src/relay/transforms/partial_eval.cc [hunk 1]
```diff
@@ -615,6 +615,8 @@ class PartialEvaluator : public ExprFunctor<PStatic(const Expr& e, LetList* ll)>
       value.push_back(ps);
       expr.push_back(ps->dynamic);
     }
+    // Note: The partial evaluator seems to do some weird stuff with sharing. Changing Tuple(expr)
+    // to WithFields(op, expr) causes some strange failures.
     return HasStatic(MkSTuple(value), ll->Push(Tuple(expr)));
```
[the remark this answers is not in the document]
Comment (electriclilies) [reply src/relay/transforms/partial_eval.cc:620]: I was seeing some failures in the partial evaluator tests, but also in other places. I'll update the comment to clarify, I don't think I need to add another unittest here though

File: src/relay/ir/expr.cc [hunk 2]
```diff
@@ -184,6 +290,25 @@ TupleGetItem::TupleGetItem(Expr tuple, int index, Span span) {
   data_ = std::move(n);
 }
 
+TupleGetItem WithFields(TupleGetItem tuple_get_item, Optional<Expr> opt_tuple,
+                        Optional<Integer> opt_index, Optional<Span> opt_span) {
+  Expr tuple = opt_tuple.value_or(tuple_get_item->tuple);
+  int index =
+      opt_index.value_or(tuple_get_item->index)
+          ->value;  // Is it OK that this is an Integer instead of int? this lets me use Optional
```
[the remark this answers is not in the document]
Comment (electriclilies) [reply src/relay/ir/expr.cc:298]: nice catch, thanks!

File: include/tvm/relay/expr.h [hunk 3]
```diff
@@ -319,8 +337,31 @@ class Call : public Expr {
...
+ * \return If all properties are null or the same as the property in the input call
+ * (i.e., opt_op is null or opt_op.value() == call->op, etc.), then we return call. Otherwise, we
+ * return a copy of call with the different fields overwritten. (i.e., if opt_op.value() !=
+ * call->op, then ret_call->op = opt_op.value()).
+ */
+Call WithFields(Call call, Optional<Expr> opt_op = Optional<Expr>(),
+                Optional<Array<Expr>> opt_args = Optional<Array<Expr>>(),
+                Optional<Attrs> opt_attrs = Optional<Attrs>(),
+                Optional<Array<Type>> opt_type_args = Optional<Array<Type>>(),
+                Optional<Span> opt_span = Optional<Span>(nullptr));
```
Comment (mbs-octoml) [inline include/tvm/relay/expr.h:363]: don't need the nullptr arg apparently

**Review by mbs-octoml — APPROVED** [review]
Look great, thank you.

For bonus points you could replace the explicit ctors in device_planner.cc with WithField.

File: src/relay/ir/expr_functor.cc [hunk 4]
```diff
@@ -512,9 +464,9 @@ class ExprBinder : public MixedModeMutator, PatternMutator {
 
   Pattern VisitPattern(const Pattern& p) final { return PatternMutator::VisitPattern(p); }
 
-  Clause VisitClause(const Clause& c) final {
-    Pattern pat = VisitPattern(c->lhs);
-    return Clause(pat, VisitExpr(c->rhs));
+  Clause VisitClause(const Clause& clause) final {
+    Pattern lhs = VisitPattern(clause->lhs);
+    return WithFields(std::move(clause), std::move(lhs), VisitExpr(clause->rhs));
```
Comment (jroesch) [inline src/relay/ir/expr_functor.cc:469]: @mbs-octoml I don't know if we actually need these moves, for const reference does this matter? I can't actually remember what the spec says about value vs move semantics for const &.

**Review by jroesch — APPROVED** [review]
Changes look great, thanks Lily! Just minor question about moving const references.

File: src/relay/ir/expr_functor.cc (hunk 4 above)
Comment (electriclilies) [reply src/relay/ir/expr_functor.cc:469]: Oh, didn't even notice it was const. I was just std::moving everything that was named that we were not using again

**Comment by electriclilies** [comment]
@mbs-octoml Changed the explicit constructions in the device planner to copy except for one Tuple, since I don't have WithFields for tuples in this PR (waiting on #9533).

**Comment by electriclilies** [comment]
@Mousius or @manupa-arm could you merge? It’s a holiday weekend here and I’d like to get this in before there are any merge conflicts. Thanks!

**Review by manupa-arm — APPROVED** [review]
LGTM. Getting this in for now.
Lets follow up if needed

**Pull request merged**